lib/automation.go

```diff
@@ -126,3 +147,183 @@ func (automationImpl) Apply(scope scope, p *ApplyParams) (*ApplyResult, error) {
+		}
+		if p.Workflow.OwnerID != wf.OwnerID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different OwnerIDs")
+		}
+	}
+	go deploy(scope, p)
+	return nil
+}
+
+func deploy(scope scope, p *DeployParams) {
+	vi := dsref.ConvertDatasetToVersionInfo(p.Dataset)
+	ref := vi.SimpleRef().String()
+	deployPayload := event.DeployEvent{
+		Ref:        ref,
+		DatasetID:  p.Dataset.ID,
+		WorkflowID: p.Workflow.ID.String(),
+	}
+	log.Debugw("deploy started", "payload", deployPayload)
+	if scope.Bus() != nil {
+		scope.Bus().PublishID(scope.Context(), event.ETDeployStart, ref, deployPayload)
+	}
+	rollback := true
+	defer func(dp *event.DeployEvent) {
+		if rollback {
+			rp := &RemoveParams{
+				Ref: ref,
+				Revision: &dsref.Rev{
+					Field: "ds",
+					Gen:   1,
+				},
+			}
+			_, err := datasetImpl{}.Remove(scope, rp)
+			if err != nil {
+				log.Debugw("deploy rollback", "error", err)
+			}
+		}
+
+	}(&deployPayload)
+
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveDatasetStart, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	saveParams := &SaveParams{
+		Ref:     vi.SimpleRef().String(),
+		Dataset: p.Dataset,
+		Apply:   false,
+	}
+	// TODO(ramfox): bandaid! remove when save can handle having saving a dataset with no body/structure
+	if p.Dataset.ID == "" && p.Dataset.BodyFile() == nil {
+		saveParams.Apply = true
+	}
+	ds, err := datasetImpl{}.Save(scope, saveParams)
+	if err != nil && !errors.Is(err, dsfs.ErrNoChanges) {
+		log.Errorw("deploy save dataset", "error", err)
+		deployPayload.Error = err.Error()
+		go func(dp *event.DeployEvent) {
+			if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+				log.Debug(err)
+			}
+		}(&deployPayload)
+		return
+	}
+
+	deployPayload.DatasetID = ds.ID
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveDatasetEnd, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	wf := p.Workflow.Copy()
+	wf.DatasetID = ds.ID
+	wf.OwnerID = scope.ActiveProfile().ID
+
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveWorkflowStart, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	wf, err = scope.Automation().SaveWorkflow(wf)
+	if err != nil {
+		log.Errorw("deploy save workflow", "error", err)
+		deployPayload.Error = err.Error()
+		go func(dp *event.DeployEvent) {
+			if err := scope.Bus().PublishID(scope.Context(), event.ETDeployEnd, ref, deployPayload); err != nil {
+				log.Debug(err)
+			}
+		}(&deployPayload)
+		return
+	}
+
+	deployPayload.WorkflowID = wf.ID.String()
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveWorkflowEnd, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	if p.Run {
```

Comment:
to avoid double-running while we work on saving without dataset bodies, let's skip running if `saveParams.Apply` is true:

```go
if p.Run && !saveParams.Apply
```